rgw/cloud-transition: add per-bucket target options

mheler opened the pull request:
Adds per-bucket targeting for S3 lifecycle transitions. Instead of transitioning all objects to a single shared destination bucket, administrators can now configure each source bucket to transition to its own dedicated destination bucket.

New Configuration Options

- target_by_bucket (boolean, default: false) - Enables per-bucket targeting. When true, each source bucket transitions to a separate destination bucket instead of sharing one.
- target_by_bucket_prefix (string, optional) - Template for destination bucket names when target_by_bucket is enabled. Supports ${zonegroup}, ${storage_class}, ${bucket} ${owner} and ${tenant} variables. Defaults to rgwx-${zonegroup}-${storage_class}-${bucket} if unset.

As part of these PR changes, there is also a helper function to determine if the remote bucket exists before attempting to create a new bucket.

Fixes: https://tracker.ceph.com/issues/73969
Fixes: https://tracker.ceph.com/issues/73144

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [X] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [X] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [X] No impact that needs to be tracked
- Documentation (select at least one)
  - [X] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [X ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

mheler committed:
rgw/cloud-transition: add per-bucket target options

Add per-bucket cloud tier targeting via new options target_by_bucket
and target_by_bucket_prefix, and use them in transition/restore to
derive the destination bucket name

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>

anthonyeleven reviewed (commented):
Docs LGTM, code approval will need to be from @cbodley et al.

mheler commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -109,10 +111,22 @@ Cloud Transition Specific Configurables
+
+.. note::
+   S3 bucket naming constraints still apply on the destination side (lowercase letters and numbers, 3-63 characters, no slashes). RGW lowercases the derived bucket name, however if you include slashes or other invalid characters in a custom template, bucket creation may still fail on the target cloud.
+
```
Not a problem, pushed those doc changes to be ~80 columns

soumyakoduri commented on src/rgw/rgw_zone.cc:
```diff
@@ -662,11 +663,70 @@ void RGWZoneGroupPlacementTierS3::decode_json(JSONObj *obj)
+    }
+  }
+
+  // these substitutions apply to both target_path and target_by_bucket_prefix
```
if this change is applied to `target_path` as well then it would affect upgrade right. Restore of the objects transitioned before upgrade will fail.

mheler replied:
If it's needed I can create another PR that leverages the object manifest for restore. This way target_path, and target_by_bucket_prefix can be changed in a storage class, without impacting transitioned objects.

mheler replied:
> > > If it's needed I can create another PR that leverages the object manifest for restore. This way target_path, and target_by_bucket_prefix can be changed in a storage class, without impacting transitioned objects.
> > 
> > 
> > we had thought about it during initial design of `restore` feature. But we decided to not to do it for couple of reasons. One of main reasons was that if there is an update to cloud endpoint credentials (which will be updated in zonegroup->placement->tier) but not in object's manifest , then that would prevent restore from happening.
> 
> That makes sense. I was thinking of simply deriving the target bucket name from the stored tier config in the object manifest rather than the current zone config. This ensures restore looks for objects where they were actually transitioned, even if target_path, target_by_bucket, or target_by_bucket_prefix have been changed in the storage class configuration since then.
> 
> Connection parameters (endpoint, credentials, region) still use current config since these are operational and may rotate.

I went ahead and added a commit to this PR for this. It should resolve the concerns around breaking cloud restore if the storage class is updated.

mheler replied:
> > > > If it's needed I can create another PR that leverages the object manifest for restore. This way target_path, and target_by_bucket_prefix can be changed in a storage class, without impacting transitioned objects.
> > > 
> > > 
> > > we had thought about it during initial design of `restore` feature. But we decided to not to do it for couple of reasons. One of main reasons was that if there is an update to cloud endpoint credentials (which will be updated in zonegroup->placement->tier) but not in object's manifest , then that would prevent restore from happening.
> > 
> > 
> > That makes sense. I was thinking of simply deriving the target bucket name from the stored tier config in the object manifest rather than the current zone config. This ensures restore looks for objects where they were actually transitioned, even if target_path, target_by_bucket, or target_by_bucket_prefix have been changed in the storage class configuration since then.
> > Connection parameters (endpoint, credentials, region) still use current config since these are operational and
> > may rotate.
> 
> I believe it is not ideal to read some parameters from the object manifest and others from the tier_config. This split can lead to confusion in the codebase and can be error prone (for example, region/location_constraint can also be tied with the target bucket). Could you please clarify the rationale for changing the current default behavior?

My concern is a user may modify the storage class to either 1) change target_path or 2) switch to using this new option. In either case, this would break the restore of that transitioned object (if it occurred previously before the storage class being modified). I didn't find any documentation showing this as a limitation (unless I missed something). I can take out the commit if it's a concern, and create a separate PR for that.

MrWowsers replied:
As an administer who is planning to use read through, I think it would be nice to support some sort of way to migrate buckets or cloud providers without user intervention. Keeping data on the object would be good to support that switch. However, it may be difficult with data on every object manifest since we don't really have tools (that I know of) to modify that data. I don't think it would be fun to modify so many objects if there are enough.

If administers who use read through wanted to migrate their cloud transition/restore to a new `target_path` or even another cloud, it may be advantageous to set that data (`target_path`, url, keys) for each bucket involved in cloud transition/restore instead of each object. If it isn't set, then use the data from the storage class.

Administers could move/migrate data using rclone or similar to a new bucket and then update the `target_path` (cloud url/keys). Having the ability to set this on a per bucket basis would ease some of the migration process to break it up in bucket chunks instead of migrating all data and changing the target_path for the whole cluster.

This would still be potentially 2 locations for storing/reading data but it would aide with migrations and allow some buckets to keep their old settings if needed.

soumyakoduri commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -50,6 +50,8 @@ Cloud Storage Class Tier Configuration
                 "dest_id": <dest_id> } ... ],
     "location_constraint": <location-constraint>,
     "target_path": <target_path>,
+    "target_by_bucket": <true | false>,
```
It might help to add an example using these options.

mheler replied:
I have included that.

soumyakoduri commented:
On a related note, (maybe in a separate PR), to address https://tracker.ceph.com/issues/73144, cloud-tier module maybe can issue `get-bucket` on the destination bucket first and only if it retunrs `ENOENT`, it may proceed with `create-bucket` request .

mheler committed:
rgw/cloud-transition: check bucket existence before create

Add HEAD request to check if target bucket exists before attempting
to create it. This avoids unnecessary PUT requests when the bucket
already exists on the remote endpoint.

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>

mheler commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -109,13 +111,45 @@ Cloud Transition Specific Configurables
-  Specifies the region where the target bucket will be created on the remote S3 endpoint. For AWS, this location needs to be specified only if the region is other than US East (us-east-1).
+  Specifies the region where the target bucket will be created on the remote
+  S3 endpoint. For AWS, specify this only if the region is other than
+  US East (us-east-1).
```
Done

mheler commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -109,13 +111,45 @@ Cloud Transition Specific Configurables
   A string that defines how the target path is constructed. The target path
   specifies a prefix to which the source bucket-name/object-name is appended.
-  If not specified the ``target_path`` created is ``rgwx-${zonegroup}-${storage-class}-cloud-bucket``.
+  If not specified and ``target_by_bucket`` is false, the ``target_path``
```
Done

soumyakoduri reviewed (commented):
Thanks.. I do not have any further comments. But I think tests can be added for these new options in `s3-tests`

soumyakoduri commented:
Thanks for the s3-tests. Maybe we could add new cloud-transition task [here](https://github.com/ceph/ceph/tree/main/qa/suites/rgw/cloud-transition/tasks) to run `target_by_bucket` specific testcases. The new options need to be parsed in [rgw_cloudtier.py](https://github.com/ceph/ceph/blob/main/qa/tasks/rgw_cloudtier.py)

mheler committed:
qa/rgw: add teuthology support for target_by_bucket cloud transition

Add cloud_target_by_bucket and cloud_target_by_bucket_prefix options
to rgw_cloudtier.py and s3tests.py. Create new test suite to run
target_by_bucket-specific s3-tests.

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>

mheler commented:
jenkins test api

mheler commented:
@anrao19 how is the testing going here?

mheler committed:
rgw/cloud-transition: yield in cloud_tier_bucket_exists HEAD

The HEAD request used null_yield, so every attempt (including the
retries added by retry_on_busy) blocked the LC worker thread for
the full HTTP timeout instead of yielding.

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>

anrao19 commented:
jenkins test make check arm64

soumyakoduri commented:
The new [s3-tests](https://github.com/ceph/s3-tests/pull/716/changes/2fc33fc980b1e5d474e463e4e48db20c2a244c65#top) added got skipped in the teuthology tests 

https://qa-proxy.ceph.com/teuthology/soumyakoduri-2026-04-27_06:12:19-rgw:cloud-transition-target_path_by_bucket-distro-default-trial/179350/teuthology.log 
```

2026-04-27T06:22:06.913 INFO:teuthology.orchestra.run.trial169.stdout:s3tests/functional/test_s3.py::test_lifecycle_cloud_transition_target_by_bucket SKIPPED [ 16%]
2026-04-27T06:22:06.968 INFO:teuthology.orchestra.run.trial169.stdout:s3tests/functional/test_s3.py::test_lifecycle_cloud_transition_target_by_bucket_multiple_buckets SKIPPED [ 33%]
```

`target_by_bucket` needs to be added and set to `true` in s3tests.conf for these tests to be run in teuthology [you may refer to `cloud_allow_read_through` option in the current code for the changes needed]. 
However I guess this can be addressed in another PR as the current changes are already being tested by QE .

mheler commented:
rebase with main

adamemerson reviewed (approved):
LGTM

soumyakoduri commented:
cloud-transition test-suite results (with new teuthology changes and s3-tests added) - https://pulpito.ceph.com/soumyakoduri-2026-05-01_15:43:06-rgw:cloud-transition-target_path_by_bucket-distro-default-trial/

anrao19 commented:
> FYI, pr testing is in progress... Tracker created for same is : https://tracker.ceph.com/issues/75122

pr testing completed and approved by @ivancich .
